Clerk Elements (Beta) - Initial docs (#890)
clerk/clerk-docs: docs/elements/guides/sign-in.mdx

---
description: Learn how to build a complete sign-in form with Clerk Elements.
---

# Build a sign-in flow

> 

Reviewer: ```suggestion

<TutorialHero
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
    },
    {
      title: "Create a Next.js application",
      link: "https://nextjs.org/docs/getting-started/installation",
    },
    {
      title: "Upgrade to Core 2 if necessary",
      link: "https://clerk.com/docs/upgrade-guides/core-2/nextjs#upgrading-clerk-nextjs-to-core-2",
    },
  ]}
>

- Build a complete sign-in flow using Clerk Elements

</TutorialHero>

```

By adding this, we can omit the opening sentence and remove some text in the `<Callout>` below while making this guide look consistent with some of our other guides that have prerequisites 👍. Here's what it looks like:

![image](https://github.com/clerk/clerk-docs/assets/5156915/184c7652-f164-46a5-9b4e-c4abfc303e2a)

> In this guide you'll learn how to build a complete sign-in flow using Clerk Elements.

Reviewer: As far as I can tell, we have no rules around this, so either way should be fine for now 👍 
- https://github.com/clerk/clerk-docs/blob/main/STYLEGUIDE.md

> In this guide you will learn how to build a complete sign-in flow using Clerk Elements.

Reviewer: ```suggestion
```
Unnecessary with above suggestion

<Callout type="info">

> Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you are starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.

Reviewer: ```suggestion
Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added.
```
Deleted some lines that are unnecessary with above suggested change

</Callout>

<Steps>

### Add a sign-in route

> Create a new route in your Next.js application. The route needs to be an [optional catch-all route](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments) so the sign-in flow can handled nested paths. Add the following snippet to `/app/sign-in/[[...sign-in]]/page.tsx`:

Reviewer: ```suggestion
Create a new route in your Next.js application. The route needs to be an [optional catch-all route](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments) so the sign-in flow can handled nested paths, as shown in the following example:
```
We should avoid repeating the file name to avoid introducing discrepancies if we ever edit this line but forget to edit the code sample (or vice versa)

```tsx filename="/app/sign-in/[[...sign-in]]/page.tsx"
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
    </SignIn.Root>
  )
}
```

> You've added two imports, which you'll use as you build out the rest of the flow. `<SignIn.Root>` manages the sign-in state and handles connecting the components to Clerk's APIs.

Reviewer 2: ```suggestion
You will use these two imports to build out the rest of the flow. `<SignIn.Root>` manages the sign-in state and handles connecting the components to Clerk's APIs.
```

### Add the start step

> The Clerk authentication flows are made up of **steps**. Steps handle rendering the UI for each part of the flow. To allow users to create a sign-in attempt, the `start` step needs to be rendered. Use the `<SignIn.Step>` component to render this:

Reviewer: ```suggestion
The Clerk authentication flows are made up of **steps**. Steps handle rendering the UI for each part of the flow. To allow users to create a sign-in attempt, the `start` step needs to be rendered. The following example does so with the `<SignIn.Step>` component:
```

```tsx filename="/app/sign-in/[[...sign-in]]/page.tsx" {9-11}
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
      <SignIn.Step name="start">
        <h1>Sign in to your account</h1>
      </SignIn.Step>
    </SignIn.Root>
  )
}
```

### Add form fields

> Make it functional by adding input fields. Utilize Clerk Elements' common components to render an `identifier` field, as well as the `<Connection>` component to allow users to sign in with a social connection, like Google:

Reviewer: ```suggestion
Make it functional by adding input fields. The following example uses the `Clerk.Fields` component to render an `identifier` field, as well as the `<Connection>` component to allow users to sign in with a social connection, like Google:
```

```tsx filename="/app/sign-in/[[...sign-in]]/page.tsx" {12-22}
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
      <SignIn.Step name="start">
        <h1>Sign in to your account</h1>

        <Clerk.Connection name="google">
          Sign in with Google
        </Clerk.Connection>

        <Clerk.Field name="identifier">
          <Clerk.Label>Email</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignIn.Action submit>Continue</SignIn.Action>
      </SignIn.Step>
    </SignIn.Root>
  )
}
```

`<Clerk.Field>` takes care of wiring up the input with the label element, and `<Clerk.FieldError>` will render any field-specific errors that get returned from Clerk's API. You've also used the `<SignIn.Action>` component. This component provides common actions that are used throughout the flows. In this case, you're using the `submit` action to render a submit button for the start form.

<Callout type="info">

> If your Clerk instance supports signing in with Google and doesn't require MFA, you should be able to complete a sign-in with the components you've rendered so far!

Reviewer 2: ```suggestion
If your Clerk instance supports signing in with Google and doesn't require MFA (multi-factor authentication), you should be able to complete a sign-in with the components you've rendered so far!
```

</Callout>

### Add verification

> As users progress through a sign-in attempt, they may be asked to **verify** a number of authentication factors. This is done in the `verifications` step. Depending on your instance configuration and the state of the sign-in attempt, certain strategies will require different fields. To support conditional rendering based on the strategy being verified, you can use the `<SignIn.Strategy>` component. Assuming that your instance is configured to accept email codes, add the code necessary to accept them:

Reviewer: ```suggestion
As users progress through a sign-in attempt, they may be asked to **verify** a number of authentication factors in the `verifications` step. You can render a form for the user to complete verification, but each [verification strategy](/docs/references/javascript/sign-up/verification#attempt-verification-params) requires different fields. It's best to render the form fields conditionally using the `<SignIn.Strategy>` component.

The following example demonstrates how to conditionally render a form for the `email_codes` strategy:
```

Breaks up the text and makes the wording a little more direct

> As one progresses through a sign-in attempt, your users may be asked to **verify** a number of authentication factors. This is done in the `verifications` step. Depending on your instance configuration and the state of the sign-in attempt, certain strategies will require different fields. To support conditional rendering based on the strategy being verified, you will use the `<SignIn.Strategy>` component. Assuming that your instance is configured to accept email codes, add the code necessary to accept them:

Reviewer 2: ```suggestion
As users progress through a sign-in attempt, they may be asked to **verify** a number of authentication factors. This is done in the `verifications` step. Depending on your instance configuration and the state of the sign-in attempt, certain strategies will require different fields. To support conditional rendering based on the strategy being verified, you can use the `<SignIn.Strategy>` component. Assuming that your instance is configured to accept email codes, add the code necessary to accept them:
```

```tsx filename="/app/sign-in/[[...sign-in]]/page.tsx" {25-38}
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
      <SignIn.Step name="start">
        <h1>Sign in to your account</h1>

        <Clerk.Connection name="google">
          Sign in with Google
        </Clerk.Connection>

        <Clerk.Field name="identifier">
          <Clerk.Label>Email</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignIn.Action submit>Continue</SignIn.Action>
      </SignIn.Step>

      <SignIn.Step name="verifications">
        <SignIn.Strategy name="email_code">
          <h1>Check your email</h1>
          <p>We sent a code to <SignIn.SafeIdentifier />.</p>

          <Clerk.Field name="code">
            <Clerk.Label>Email code</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignIn.Action submit>Continue</SignIn.Action>
        </SignIn.Strategy>
      </SignIn.Step>
    </SignIn.Root>
  )
}
```

> In the verifications step, you must render a `<SignIn.Strategy>` component for each authentication strategy that your instance supports. This will also handle MFA if it is enabled for your instance. Check out the [Sign-in reference](/docs/elements/reference/sign-in#strategy) for a full list of possible authentication strategies.

Reviewer: ```suggestion
```

Not necessary with above suggested change

> In the verifications step, you will need to render a `<SignIn.Strategy>` component for each authentication strategy that your instance supports. This will also handle multi-factor authentication (MFA) if it is enabled for your instance. Check out the [Sign-in reference](/docs/elements/reference/sign-in#strategy) for a full list of possible authentication strategies.

Reviewer: ```suggestion
In the verifications step, you must render a `<SignIn.Strategy>` component for each authentication strategy that your instance supports. This will also handle MFA if it is enabled for your instance. Check out the [Sign-in reference](/docs/elements/reference/sign-in#strategy) for a full list of possible authentication strategies.
```

Verification is the final step in the sign-in flow. When a user has verified all required factors, the sign-in attempt will be complete and they will be signed in.

### Add password support

> If your instance is configured to support authenticating with passwords, you will need to add a few additional steps and verification strategies. You can choose if you want to support providing a password in the start step with an additional field, or as an additional verification strategy. For this guide, add it as a standalone verification strategy.

Reviewer: ```suggestion
If your instance is configured to support authenticating with passwords, you will must add a few additional steps and verification strategies. You can choose if you want to support providing a password in the `start` step with an additional field, or as an additional verification strategy. For this guide, add it as a standalone verification strategy.
```

> If your instance is configured to support authenticating with passwords, you'll need to add a few additional steps and verification strategies. You can choose if you want to support providing a password in the start step with an additional field, or as an additional verification strategy. For this guide, add it as a standalone verification strategy.

Reviewer 2: ```suggestion
If your instance is configured to support authenticating with passwords, you will need to add a few additional steps and verification strategies. You can choose if you want to support providing a password in the start step with an additional field, or as an additional verification strategy. For this guide, add it as a standalone verification strategy.
```

```tsx filename="/app/sign-in/[[...sign-in]]/page.tsx" {39-65,68-94}
'use client'

import * as Clerk from '@clerk/elements/common'
import * as SignIn from '@clerk/elements/sign-in'

export default function SignInPage() {
  return (
    <SignIn.Root>
      <SignIn.Step name="start">
        <h1>Sign in to your account</h1>

        <Clerk.Connection name="google">
          Sign in with Google
        </Clerk.Connection>

        <Clerk.Field name="identifier">
          <Clerk.Label>Email</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignIn.Action submit>Continue</SignIn.Action>
      </SignIn.Step>

      <SignIn.Step name="verifications">
        <SignIn.Strategy name="email_code">
          <h1>Check your email</h1>

>           <p>We've sent a code to <SignIn.SafeIdentifier />.</p>

Reviewer 2: ```suggestion
          <p>We sent a code to <SignIn.SafeIdentifier />.</p>
```

          <Clerk.Field name="code">
            <Clerk.Label>Email code</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignIn.Action submit>Continue</SignIn.Action>
        </SignIn.Strategy>

        <SignIn.Strategy name="password">
          <h1>Enter your password</h1>

          <Clerk.Field name="password">
            <Clerk.Label>Password</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignIn.Action submit>Continue</SignIn.Action>
          <SignIn.Action navigate="forgot-password">
            Forgot password?
          </SignIn.Action>
        </SignIn.Strategy>

        <SignIn.Strategy name="reset_password_email_code">
          <h1>Check your email</h1>

>           <p>We've sent a code to <SignIn.SafeIdentifier />.</p>

Reviewer 2: ```suggestion
          <p>We sent a code to <SignIn.SafeIdentifier />.</p>
```

          <Clerk.Field name="code">
            <Clerk.Label>Email code</Clerk.Label>
            <Clerk.Input />
            <Clerk.FieldError />
          </Clerk.Field>

          <SignIn.Action submit>Continue</SignIn.Action>
        </SignIn.Strategy>
      </SignIn.Step>

      <SignIn.Step name="forgot-password">
        <h1>Forgot your password?</h1>

        <SignIn.SupportedStrategy name="reset_password_email_code">
          Reset password
        </SignIn.SupportedStrategy>

        <SignIn.Action navigate="back">Go back</SignIn.Action>
      </SignIn.Step>

      <SignIn.Step name="reset-password">
        <h1>Reset your password</h1>

        <Clerk.Field name="password">
          <Clerk.Label>New password</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <Clerk.Field name="confirmPassword">
          <Clerk.Label>Confirm password</Clerk.Label>
          <Clerk.Input />
          <Clerk.FieldError />
        </Clerk.Field>

        <SignIn.Action submit>Reset password</SignIn.Action>
      </SignIn.Step>
    </SignIn.Root>
  )
}
```

> To enable users to reset their passwords, two additional steps, `forgot-password` and `reset-password`, can be added. The `forgot-password` step initiates the reset process, whereby an email code is sent to the user for verification. Once verified, the `reset-password` step allows the user to input a new password. If your instance has been set up to accept SMS codes, the `reset_password_phone_code` strategy can also be used.

Reviewer: ```suggestion
To enable users to reset their passwords, you can add the following additional steps:

1. `forgot-password` –  Initiates the reset process, whereby an email code is sent to the user for verification.
1. `reset-password` – Allows a verified user to input a new password. If your instance has been set up to accept SMS codes, you can also use `reset_password_phone_code`.
```

Makes it a little easier to read

Reviewer: (Also, markdown auto-numbers ordered lists, so it's better to make each item 1)

> When accepting passwords, you also need a way to let users reset them. To support this, you've added two new steps: `forgot-password` and `reset-password`. The `forgot-password` step lets your user trigger the reset flow. First, it sends an email code that needs to be verified, secondly the `reset-password` step then accepts a new password upon successful verification. If your instance is configured to accept SMS codes, the `reset_password_phone_code` strategy can also be used.

Reviewer 2: ```suggestion
To enable users to reset their passwords, two additional steps, `﻿forgot-password` and ﻿`reset-password`, can be added. The ﻿`forgot-password` step initiates the reset process, whereby an email code is sent to the user for verification. Once verified, the ﻿`reset-password` step allows the user to input a new password. If your instance has been set up to accept SMS codes, the ﻿`reset_password_phone_code` strategy can also be used.
```

> The `forgot-password` step renders a new component, [`<SignIn.SupportedStrategy>`](/docs/elements/reference/sign-in#supportedstrategy). This component is also used in the `forgot-password` and `choose-strategy` steps to trigger the verification of a supported strategy. In this context, it triggers the verification of one of the reset_password strategies.

Reviewer: ```suggestion
```
Not necessary with above changes

> In the `forgot-password` step, you'll notice that a new component is rendered, [`<SignIn.SupportedStrategy>`](/docs/elements/reference/sign-in#supportedstrategy). In the `forgot-password` and `choose-strategy` steps, `<SignIn.SupportedStrategy>` is used to trigger a verification of a supported strategy. In this case, you're using it to trigger verification of one of the `reset_password` strategies.

Reviewer 2: ```suggestion
The `﻿forgot-password` step renders a new component, [﻿<SignIn.SupportedStrategy>](https://github.com/docs/elements/reference/sign-in#supportedstrategy). This component is also used in the `﻿forgot-password` and ﻿`choose-strategy` steps to trigger the verification of a supported strategy. In this context, it triggers the verification of one of the ﻿reset_password strategies.
```

<Callout>

>   If your instance isn't configured to use passwords, or any of the strategies outlined here, Clerk Elements will log a warning to the console during development.

Reviewer: ```suggestion
  If your instance isn't configured to use any of the strategies outlined here, Clerk Elements will log a warning to the console for development instances.
```
I think this is more clear, but is my change accurate? If no we should reword so it's clearer what "in development" means

Reviewer 3: "In development" here means when `NODE_ENV` is set to `development`. While that information is correct, it's not really helpful for the end user. In reality, it means: When you run the development server of your framework, e.g. `next dev`

</Callout>

### Customize and add styling

> You may have noticed that none of the components displayed to this point include styling. This is intentional! With the exception of `<SignIn.Root>` and `<SignIn.Strategy>`, every component rendered so far can be styled. The `className` prop enables the addition of custom classes to each component. Check out the [styling guide](/docs/elements/guides/styling) for a more in-depth guide on styling Clerk Elements.

Reviewer: ```suggestion
Learn how to style your Clerk Elements components with the [styling guide](/docs/elements/guides/styling).
```

> You might have noticed by now that none of components rendered so far come with their own styling. This is by design! With the exception of `<SignIn.Root>` and `<SignIn.Strategy>`, every component rendered so far can be styled. You can use the `className` prop to add custom classes to each component. Check out the [styling guide](/docs/elements/guides/styling) for a more in-depth guide on styling Clerk Elements.

Reviewer 2: ```suggestion
You may have noticed that none of the components displayed to this point include styling. This is intentional! With the exception of `<SignIn.Root>` and `<SignIn.Strategy>`, every component rendered so far can be styled. The ﻿`className` prop enables the addition of custom classes to each component. Check out the [styling guide](/docs/elements/guides/styling) for a more in-depth guide on styling Clerk Elements.
```

> If you're looking to customize the UI even further, check out additional Clerk Elements components like [`<Loading>`](/docs/elements/reference/common#loading) and [`<FieldState>`](/docs/elements/reference/common#fieldstate).

Reviewer 2: ```suggestion
For more extensive customization of the UI, check out additional Clerk Elements components such as [`<Loading>`](/docs/elements/reference/common#loading) and [`<FieldState>`](/docs/elements/reference/common#fieldstate).
```

---
description: Learn how to build a complete sign-in form with Clerk Elements.
---

# Build a sign-in flow with Clerk Elements

<TutorialHero
  beforeYouStart={[
    {
      title: "Follow the Next.js quickstart",
      link: "https://clerk.com/docs/quickstarts/nextjs",
    },
    {
      title: "Upgrade to Core 2 (if necessary)",
      link: "https://clerk.com/docs/upgrade-guides/core-2/nextjs",
    },
  ]}
>

- Build out different steps of an authentication flow
- Accept user input through form fields
- Verify a user's identity during a sign-in attempt
- Support various password-related features

</TutorialHero>

<Callout type="info">

- Clerk Elements is for [advanced use-cases](/docs/elements/overview#why-use-clerk-elements) that require a high-level of customization. The easiest way to implement Clerk is with our [all-in-one UI components](/docs/components/overview).

- Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added.

</Callout>

<Steps>

### Add a sign-in route

> Create a new route in your Next.js application. The route needs to be an [optional catch-all route](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments) so the sign-in flow can handled nested paths, as shown in the following example:

Reviewer 4: We removed it because we shouldn't have repetition like that in prose - the codeblocks were recently redesigned so that the filename and copy button can stand out more so hopefully that should help users more with seeing the filename!
